### PR TITLE
Draw radar update line for mobile sensors array (TS)

### PR DESCRIPTION
**Pull request by teees**
http://gfycat.com/VioletQuarterlyAlabamamapturtle
EDIT:
- Added contrast http://gfycat.com/FancySickKiwi
- Added trailing lines http://gfycat.com/AcademicQualifiedAssassinbug
- Made the colors user definable and polished the comments
- Added contrast to trailing lines to make them more visible on snow maps http://gfycat.com/AchingRegularBrocketdeer

### CODE REVIEW

**Comment by Phrohdoh** [comment]
We definitely want the "non-gappy" circle for this, but looks great!

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 1]
```diff
@@ -19,14 +19,29 @@ namespace OpenRA.Mods.Common.Traits
 {
 	class RenderDetectionCircleInfo : ITraitInfo, Requires<DetectCloakedInfo>
 	{
-		public object Create(ActorInitializer init) { return new RenderDetectionCircle(init.Self); }
+		[Desc("Draw radar update line")]
+		public readonly bool DrawUpdateLine = false;
+
+		[Desc("Radar update line tick")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: Please explain the default value.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 2]
```diff
@@ -19,14 +19,29 @@ namespace OpenRA.Mods.Common.Traits
 {
 	class RenderDetectionCircleInfo : ITraitInfo, Requires<DetectCloakedInfo>
 	{
-		public object Create(ActorInitializer init) { return new RenderDetectionCircle(init.Self); }
+		[Desc("Draw radar update line")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: `Draw line from this actor to the detection circle` or similar please, these descriptions don't offer anything over the field names as-is.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 3]
```diff
@@ -47,6 +62,34 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr)
 				0,
 				Color.FromArgb(128, Color.LimeGreen),
 				Color.FromArgb(96, Color.Black));
+
+			if (info.DrawUpdateLine)
+			{
+				for (int i = info.LineTrailLength; i >= 0; i--)
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: `var`

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 4]
```diff
@@ -47,6 +62,34 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr)
...
+							length,
+							3,
+							Color.FromArgb(96, Color.Black));
+					var alpha = 128 - (info.LineTrailLength > 0 ? (int)(i * 128f / info.LineTrailLength) : 0);
+					yield return new BeamRenderable(
+						self.CenterPosition,
+						0,
+						length,
+						1,
+						Color.FromArgb(alpha, Color.LimeGreen));
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: Please pull this out into the Info class.

**Comment by teees** [comment]
Let me know when you think it looks fine, will then rebase that last polish commit

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 5]
```diff
@@ -19,14 +19,35 @@ namespace OpenRA.Mods.Common.Traits
 {
 	class RenderDetectionCircleInfo : ITraitInfo, Requires<DetectCloakedInfo>
 	{
-		public object Create(ActorInitializer init) { return new RenderDetectionCircle(init.Self); }
+		[Desc("Draw a rotating radar scanner update line, disabled by default.")]
+		public readonly bool DrawUpdateLine = false;
+
+		[Desc("Number of (angle / 1024) the Radar update line advances per tick.")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: This still doesn't explain what `-1` means in this context.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: It's a negative angle - which results with a counterclockwise animation. It's not a magic value.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: That'd go great in the description.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: There's nothing special about a negative angle, though.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 6]
```diff
@@ -19,14 +19,35 @@ namespace OpenRA.Mods.Common.Traits
...
+		public readonly int UpdateLineTick = -1;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: This should be a WAngle for consistency in the docs.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 7]
```diff
@@ -19,14 +19,35 @@ namespace OpenRA.Mods.Common.Traits
...
+
+		public object Create(ActorInitializer init) { return new RenderDetectionCircle(init.Self, this); }
 	}
 
-	class RenderDetectionCircle : IPostRenderSelection
+	class RenderDetectionCircle : ITick, IPostRenderSelection
 	{
+		readonly RenderDetectionCircleInfo info;
 		readonly Actor self;
+		int lineAngle;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: This probably wants to be a `WAngle` too.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 8]
```diff
@@ -45,8 +66,36 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr)
...
-				Color.FromArgb(96, Color.Black));
+				info.Color,
+				info.ContrastColor);
+
+			if (info.DrawUpdateLine)
+			{
+				for (var i = info.LineTrailLength; i >= 0; i--)
+				{
+					var angle = new WAngle((lineAngle - i * info.UpdateLineTick) % 1024);
+					var length = new WVec((int)(range.Length * (angle.Cos() / 1024f)), (int)(range.Length * (angle.Sin() / 1024f)), 0);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: You should be able to simplify this to `var length = range.Length * new WVec(angle.Cos(), angle.Sin(), 0) / 1024;`

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 9]
```diff
@@ -45,8 +66,36 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr)
...
+					if (i == 0)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: This only draws the contrast color on the main beam, which then leaves the trails almost invisible on snow maps.  Seems a bit unpolished.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs [hunk 10]
```diff
@@ -45,8 +66,36 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr)
...
+						yield return new BeamRenderable(
+							self.CenterPosition,
+							0,
+							length,
+							3,
+							info.ContrastColor);
+					var alpha = info.Color.A - (info.LineTrailLength > 0 ? (int)(i * info.Color.A / (float)info.LineTrailLength) : 0);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: Likewise here you don't need to cast to a float and then back.  Integer multiplication and division does what you want.

**Comment by pchote** [comment]
Looks good overall, but see above code + polish nits.

File: OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs (hunk 9 above)
Comment (teees) [inline OpenRA.Mods.Common/Traits/Render/RenderDetectionCircle.cs]: That was intentionally done back then, changed it.

**Comment by teees** [comment]
polished and rebased

**Comment by abcdefg30** [comment]
Works ingame. :+1: 
detectioncircle
Although it looks a bit weird if I want the circle to circle faster:
detectioncircleb

**Comment by teees** [comment]
I just noticed there are some issues with the transition to WAngle I need to sort out before this can be merged

**Comment by teees** [comment]
Solved, I wondered why `Math.Sign(WAngle.Angle)` wasn't working for me ... turns out WAngle is always converted to a positive angle internally

**Comment by abcdefg30** [comment]
Looks better now. :+1: still stands.

**Comment by Mailaender** [comment]
Nicely done! :+1:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/993d5faef6f8d70c0492f2c4997f86599629c5af...8d441ec9d2a08bf0e85a1815fc9dee8bff393779)

**Pull request merged**